File: androidEats/src/main/java/com/dunk/eats/android/presentation/recipe_detail/RecipeDetailViewModel.kt
package com.dunk.eats.android.presentation.recipe_detail

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dunk.eats.domain.model.ErrorMessage
import com.dunk.eats.domain.model.PositiveAction
import com.dunk.eats.domain.model.UIComponentType
import com.dunk.eats.domain.util.ErrorMessageQueueUtil
import com.dunk.eats.domain.util.Queue
import com.dunk.eats.interactors.recipe_detail.GetRecipe
import com.dunk.eats.presentation.recipe_detail.RecipeDetailEvents
import com.dunk.eats.presentation.recipe_detail.RecipeDetailState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.*
import javax.inject.Inject

@ExperimentalStdlibApi
@HiltViewModel
class RecipeDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val getRecipe: GetRecipe
) : ViewModel() {

    val state: MutableState<RecipeDetailState> = mutableStateOf(RecipeDetailState())

    init {
        savedStateHandle.get<Int>("recipeId")?.let { recipeId ->
            onTriggerEvent(RecipeDetailEvents.GetRecipe(recipeId))
        }
    }

    fun onTriggerEvent(event: RecipeDetailEvents){
         when(event){
             is RecipeDetailEvents.GetRecipe -> {
                 getRecipe(recipeId = event.recipeId)
             }
             is RecipeDetailEvents.RemoveHeadMessageFromQueue -> {
                 removeMessageAtHead()
             }
             else -> {
                 addErrorToQueue(
                     ErrorMessage.Builder()
                         .id(UUID.randomUUID().toString())
                         .title("Error")
                         .uiComponentType(UIComponentType.Dialog)
                         .description("Invalid event")
                         .positive(PositiveAction(positiveBtnTxt = "Ok", onPositiveAction = {}))
                 )
             }
         }
    }

    private fun getRecipe(recipeId:Int){
        getRecipe.execute(
            recipeId = recipeId
        ).collectFlow(viewModelScope) { dataState ->
            state.value = state.value.copy(dataState.isLoading)
            dataState.data?.let {
                state.value = state.value.copy(recipe = it)
            }

            dataState.errorMessage?.let { message ->
                addErrorToQueue(
                    ErrorMessage.Builder()
                        .id(UUID.randomUUID().toString())
                        .title(message.title)
                        .uiComponentType(UIComponentType.Dialog)
                        .description(message.description ?: "Unknown error")
                        .positive(PositiveAction(positiveBtnTxt = "Ok", onPositiveAction = {}))
                )
            }

        }
    }

    private fun addErrorToQueue(error: ErrorMessage.Builder) {
        if (!ErrorMessageQueueUtil().isErrorUnique(
                queue = state.value.errorQueue,
                message = error.build()
            )) {
            val queue = state.value.errorQueue
            queue.add(error.build())
            state.value = state.value.copy(errorQueue = queue)
        }
    }

    private fun removeMessageAtHead() {
        try {
            val queue = state.value.errorQueue
            queue.remove()
            // Forcing the app to recompose and remove the dialog since the queue is empty
            state.value = state.value.copy(errorQueue = Queue(mutableListOf()))
            state.value = state.value.copy(errorQueue = queue)
        }catch (e: Exception){
            // Nothing to remove queue is empty
        }
    }

}

File: iosEats/iosEats/presentation/recipe_detail/RecipeDetailViewModel.swift
//
//  RecipeDetailViewModel.swift
//  iosEats
//
//  Created by Duncan K on 27/11/2022.
//  Copyright © 2022 orgName. All rights reserved.
//


import SwiftUI
import shared

class RecipeDetailViewModel: ObservableObject {
    
    private let getRecipe: GetRecipe
    
    @Published var state: RecipeDetailState = RecipeDetailState()
    
    init(
        recipeId: Int,
        getRecipe: GetRecipe
    ) {
        self.getRecipe = getRecipe
        onTriggerEvent(stateEvent: RecipeDetailEvents.GetRecipe(recipeId: Int32(recipeId)))
    }
    
    func onTriggerEvent(stateEvent: RecipeDetailEvents){
        switch stateEvent {
            case is RecipeDetailEvents.GetRecipe:
                getRecipe(recipeId: Int((stateEvent as! RecipeDetailEvents.GetRecipe).recipeId))
            case is RecipeDetailEvents.RemoveHeadMessageFromQueue:
                removeHeadFromQueue()
            default: doNothing()
        }
    }
        
    private func getRecipe(recipeId: Int){
        do{
            self.updateState(isLoading: true)
            try self.getRecipe.execute(
                recipeId: Int32(recipeId)
            ).collectFlow(
                coroutineScope: nil,
                callback: { dataState in
                if dataState != nil {
                    let data = dataState?.data
                    let message = dataState?.errorMessage
                    let loading = dataState?.isLoading ?? false
                    self.updateState(isLoading: loading)
                    
                    if(data != nil){
                        self.updateState(recipe: data! as Recipe)
                    }
                    if(message != nil){
                        self.handleMessageByUIComponentType(message! )
                    }
                }else{
//                    self.logger.log(msg: "GetRecipe: DataState is nil")
                }
            })
        }catch{
//            logger.log(msg: "GetRecipe: ERROR: \(error)")
        }
    }
    
    private func handleMessageByUIComponentType(_ message: ErrorMessage){
        switch message.uiComponentType{
        case UIComponentType.Dialog():
            appendToQueue(message: message)
        case UIComponentType.None():
            doNothing()
//            logger.log(msg: "\(message.description)")
        default:
            doNothing()
        }
    }
    
    private func appendToQueue(message: ErrorMessage){
        let currentState = (self.state.copy() as! RecipeDetailState)
        let queue = currentState.errorQueue
        let queueUtil = ErrorMessageQueueUtil() // prevent duplicates
        if !queueUtil.isErrorUnique(queue: queue, message: message) {
           queue.add(element: message)
           updateState(queue: queue)
        }
   }

    private func removeHeadFromQueue(){
        let currentState = (self.state.copy() as! RecipeDetailState)
        let queue = currentState.errorQueue
        do {
            try queue.remove()
            updateState(queue: queue)
        }catch{
//            logger.log(msg: "\(error)")
        }
    }
    
    private func doNothing(){}
    
    private func updateState(
        isLoading: Bool? = nil,
        recipe: Recipe? = nil,
        queue: Queue<ErrorMessage>? = nil
    ){
        let currentState = (self.state.copy() as! RecipeDetailState)
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            recipe: recipe ?? currentState.recipe,
            errorQueue: queue ?? currentState.errorQueue
        )
    }
    
}
